### DISCUSSION TITLE
Tailwind 2.0 with Sage 9

### DISCUSSION THREAD

**Discussion by minemindmedia** [tailwindlabs/tailwindcss#2941] [Help] [answered]
I would love to get Tailwind 2.0 running with Sage 9 (https://roots.io)

After upgrading to 2.0 within the Sage theme via:

npm install tailwindcss@latest postcss@latest autoprefixer@latest

and using Node 12.
yarn && yarn build errors:

```
 ERROR  Failed to compile with 2 errors                                                                                                                                   

 error  in ./resources/assets/styles/main.scss

Module build failed: ModuleBuildError: Module build failed: ValidationError: Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'plugins'. These properties are valid:
   object { postcssOptions?, execute?, sourceMap? }
    at validate (/Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.loader (/Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/postcss-loader/dist/index.js:43:29)
    at /Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/webpack/lib/NormalModule.js:195:19
    at /Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/loader-runner/lib/LoaderRunner.js:233:18

 @ multi ./scripts/main.js ./styles/main.scss

 error  in ./resources/assets/styles/main.scss

Module build failed: ModuleBuildError: Module build failed: ValidationError: Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'plugins'. These properties are valid:
   object { postcssOptions?, execute?, sourceMap? }
    at validate (/Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.loader (/Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/postcss-loader/dist/index.js:43:29)
    at /Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/webpack/lib/NormalModule.js:195:19
    at /Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /Volumes/LaCie/_development/_trellis/_hybrid/trellis.project.build/site/web/app/themes/sage/node_modules/loader-runner/lib/LoaderRunner.js:233:18
```

The last version of Tailwind I used successfully with Sage was version 1.9.5
What may be the issue here with these files:

main.scsss --

```
/**
 * This injects Tailwind's base styles, which is a combination of
 * Normalize.css and some additional base styles.
 */
@tailwind base;

/** Import variables */
@import "common/variables";

/** Import everything from autoload */
@import "./autoload/**/*";

/**
 * Import npm dependencies
 *
 * Prefix your imports with `~` to grab from node_modules/
 * @see https://github.com/webpack-contrib/sass-loader#imports
 */
// @import "~some-node-module";

/**
 * This injects any component classes registered by Tailwind plugins.
 */
@tailwind components;

/** Import theme styles */
@import "components/buttons";
@import "components/comments";
@import "components/forms";
@import "components/wp-classes";
@import "layouts/header";
@import "layouts/sidebar";
@import "layouts/footer";
@import "layouts/pages";
@import "layouts/posts";
@import "layouts/tinymce";

/**
 * This injects all of Tailwind's utility classes, generated based on your
 * config file.
 */
@tailwind utilities;

```

main.js --

```
// import external dependencies
import 'jquery';

// Import everything from autoload
import './autoload/**/*'

// import local dependencies
import Router from './util/Router';
import common from './routes/common';
import home from './routes/home';
import aboutUs from './routes/about';

/** Populate Router instance with DOM routes */
const routes = new Router({
  // All pages
  common,
  // Home page
  home,
  // About Us page, note the change from about-us to aboutUs.
  aboutUs,
});

// Load Events
jQuery(document).ready(() => routes.loadEvents());

```

**Comment by simonswiss** [comment]
Hello!

Using the [PostCSS v7 `compat` version of Tailwind](https://tailwindcss.com/docs/installation#post-css-7-compatibility-build) might fix this issue for you.

**Reply by minemindmedia** [reply]
Hello!

Thank you, however, that didn't help at all.
I'll just be waiting until the Roots team makes this more easily available.

**Reply by simonswiss** [reply]
See the response below - it's a really quick fix, and I have also submitted a PR to the Sage installer repo that does the same thing 👍

**Comment by simonswiss** [comment, marked as answer]
Hey @minemindmedia,

So... I did some digging, and took Sage 9 for a spin.

I created a brand new Sage 9 project with:

```sh
composer create-project roots/sage your-theme-name
```

Turns out I had the exact same error out of the box, with Tailwind v1 (`PostCSS Loader has been initialized using an options object that does not match the API schema.`).

I found out that the `resources/assets/build/webpack.config.preset.js` file was the culprit, since it uses a deprecated way of passing PostCSS options.

## Here's how you can fix it

### 1. Update the postcss options object syntax

In that `webpack.config.preset.js` file, find the `postcss` loader and wrap the contents of the `options` object inside a `postcssOptions` object:

```diff
{
  loader: "postcss",
  options: {
+    postcssOptions: {
      parser: config.enabled.optimize
        ? "postcss-safe-parser"
        : undefined,
      plugins: postcssPlugins,
      sourceMap: false,
+    },
  },
},
``` 

This will fix that error!

### 2. Update the deprecated Tailwind 1 classnames used in `@apply` directives

Once you do that, you should be only steps away from running with Tailwind 2. You will get a couple of warnings for non-existing classnames used in `@apply` rules:

1. `shadow-outline`, used in the `_comments.scss` component, should be replaced with `ring`
2. `whitespace-no-wrap` should be replaced with `whitespace-nowrap`

There may be more in your case based on what you did with your project, but based on a new project as of today, these were the only two things to dix.

### That's it!

Once you do that, you should be up and running properly with Tailwind 2.0 on Sage 9 🎉  

![image](https://user-images.githubusercontent.com/485747/100946119-c6944f00-3556-11eb-8ee3-13a66d6546ef.png)

Hope it helps 😁

**Reply by simonswiss** [reply]
Update - since I went through these steps thoroughly, I thought I may as well open a PR on the `Sage Installer` repo so this saves people the trouble 👍

https://github.com/roots/sage-installer/pull/40

**Comment by knowler** [comment]
@minemindmedia I merged and released @simonswiss’ PR to upgrade the Tailwind preset in the Sage 9 installer. If you want to use the installer, you can do the following until we tag a release in Sage (which bumps that dep):

```bash
# Get Sage 9 either through the installer or just cloning the Sage repo at the latest 9.x branch

# Update Sage Installer to 1.6.4
$ composer update roots/sage-installer

# Run the preset (select Tailwind + yes to override files — if all your changes are stashed or committed!)
$ ./vendor/bin/sage preset
```

**Reply by simonswiss** [reply]
Awesome - thanks for following up in this thread, @knowler 🙏

**Comment by firhatsungkar** [comment]
Hi, @simonswiss I think you forgot to add postcss in dependencies on the installer. I got an error `PostCSS plugin autoprefixer requires PostCSS 8.`. And after I add those packages it runs build well.

**Reply by firhatsungkar** [reply]
I think this PR related to that issue https://github.com/roots/sage-installer/pull/41

**Reply by knowler** [reply]
(I posted this, then GitHub loaded your comment. 😛)

Hey @firhatsungkar, I have a follow-up PR at roots/sage-installer#41. I just haven’t merged it because it means we might potentially have to also increase the Node version requirement in Sage 9 and I just haven't gotten to dealing with all that.

**Comment by lewsea** [comment]
Hi, this might be a very dumb question, but here it goes
i tried running, yarn add tailwindcss@latest postcss@latest autoprefixer@latest to upgrade tailwind, then replace the postcssoptions
but when i run yarn build, yarn start, all of the tailwind classes and the @apply styling  is not recognized, and the browser console says like unknown property ? what could be the problem here ? Thanks ! [ sorry for this kind of question ]

**Reply by lewsea** [reply]
also when i run yarn build the main.css is only  5kb it looks like the @tailwind directives is not recognized.
